ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3BaseHandler.java

```diff
@@ -135,4 +148,70 @@ public static boolean isMultipartListPartRequest(RestRequest restRequest) {
     return restRequest.getRestMethod() == RestMethod.GET && restRequest.getArgs().containsKey(S3_REQUEST)
         && restRequest.getArgs().containsKey(UPLOAD_ID_QUERY_PARAM);
   }
+
+  protected class DefaultCallbackChain {
+    private final RestRequest restRequest;
+    private final Callback<ReadableStreamChannel> finalCallback;
+    private final String uri;
+    private final FrontendMetrics metrics;
+    private AccountAndContainerInjector accountAndContainerInjector;
+    private SecurityService securityService;
+    private RestServiceErrorCode returnCode;
+    private String returnMessage;
+
+    /**
+     * @param restRequest the {@link RestRequest}.
+     * @param securityService the {@link SecurityService}.
+     * @param metrics the {@link FrontendMetrics}.
+     * @param accountAndContainerInjector the {@link AccountAndContainerInjector}.
+     * @param finalCallback the {@link Callback} to call on completion.
+     * @param returnCode the {@link RestServiceErrorCode}
+     * @param returnMessage the error message to return
+     */
+    protected DefaultCallbackChain(RestRequest restRequest, SecurityService securityService, FrontendMetrics metrics,
+        AccountAndContainerInjector accountAndContainerInjector, Callback<ReadableStreamChannel> finalCallback,
+        String returnMessage, RestServiceErrorCode returnCode) {
+      this.restRequest = restRequest;
+      this.metrics = metrics;
+      this.accountAndContainerInjector = accountAndContainerInjector;
+      this.securityService = securityService;
+      this.finalCallback = finalCallback;
+      this.returnCode = returnCode;
+      this.returnMessage = returnMessage;
+      this.uri = restRequest.getUri();
+    }
+
+    /**
+     * Start the chain by calling {@link SecurityService#processRequest}.
+     */
+    protected void start() {
+      try {
+        accountAndContainerInjector.injectAccountContainerForNamedBlob(restRequest, metrics.postBlobMetricsGroup);
+        securityService.processRequest(restRequest, securityProcessRequestCallback());
+      } catch (Exception e) {
```

Comment:
Catch clause shouldn't be needed as it's already handled in BaseS3Handler.handle()

Reply:
Some caller will call DefaultCallbackChain directly like below in S3GetHandler.java. It didn't go through BaseS3Handler.handle()

```
  protected void doHandle(RestRequest restRequest, RestResponseChannel restResponseChannel,
      Callback<ReadableStreamChannel> callback) throws RestServiceException {
    RequestPath requestPath = (RequestPath) restRequest.getArgs().get(REQUEST_PATH);
    if (S3BaseHandler.isGetObjectLockConfigurationRequest(restRequest)) {
      // We don't support Bucket Object Configuration. So always return NOT_FOUND
      new DefaultCallbackChain(restRequest, securityService, frontendMetrics, accountAndContainerInjector, callback,
          "Object Lock configuration is not set.", RestServiceErrorCode.NotFound).start();
    } else if (S3BaseHandler.isListObjectRequest(restRequest)) {
      ...
    }
```
  }